### PR TITLE
Don't mark index as traceable, and other improvements

### PR DESCRIPTION
**Pull request by ezyang**
Signed-off-by: Edward Z. Yang <ezyang@fb.com>

### CODE REVIEW

**Review by zdevito — APPROVED** [review]
These seem fine. Are the CI errors real?

File: tools/autograd/gen_variable_type.py [hunk 1]
```diff
@@ -67,6 +67,18 @@
 auto${ref} ${arg_name}_ = unpack${suffix}(${arg_name}, "${arg_name}", ${arg_pos});""")
 
 FUNCTION_DECLARATION = CodeTemplate("""\
+struct ${op} : public Function {
+  using Function::Function;
```
Comment (apaszke) [inline tools/autograd/gen_variable_type.py]: There's a lot of repetition between these two snippets. Maybe jut make the superclass a template parameter and generate it somewhere in the code?

File: torch/csrc/jit/tracer.h [hunk 2]
```diff
@@ -219,8 +231,10 @@ namespace detail {
 
 // Exit code shared between exit and TraceExitHook::run
 inline void _exit(const std::shared_ptr<TracingState>& state, const variable_list& outputs) {
+  size_t i = 0;
```
Comment (apaszke) [inline torch/csrc/jit/tracer.h]: ugh I really hate these range based loops with outer counters 😕

File: test/test_jit.py [hunk 3]
```diff
@@ -959,6 +979,25 @@ def fn(x, y):
         self.assertTrue("hits: 100" in info_str)
         self.assertTrue("stage 1" in info_str)
 
+    def test_inplace_copy(self):
+        x = Variable(torch.randn(4, 4), requires_grad=True)
+        def f(x):
+            out = Variable(torch.zeros(x.size()))
+            out.copy_(x)
+            return out
```
[the remark this answers is not in the document]
Comment (ezyang) [reply test/test_jit.py]: Oh yeah, this PR was between me and Zach so I forgot to post some text.

So basically, the problem is that the implementation of backwards for view functions does exactly this idiom: it first creates a zero tensor, and then writes in the gradient inplace into the portion appropriate for the original view. So all of these functions are not traceable when the tracer can't handle this.

OTOH, I feel that the tracer *should* be able to handle this idiom, esp now that inplace updates on views are supported. Those won't be traced correctly at all today.

File: torch/csrc/jit/tracer.h [hunk 4]
```diff
@@ -163,6 +158,23 @@ inline Value* getValueTrace(const std::shared_ptr<TracingState>& state, const Va
...
+    Node *n = state->graph->createUndefined();
+    return state->graph->appendNode(n)->output();
+  }
+
+  auto vts = detail::getValueState(state, var, false);
+  if (!vts) {
+    std::ostringstream os;
+    os << "output " << output_no << " of traced region did not have observable "
+       << "data dependence with trace inputs; this probably indicates your program "
+       << "cannot be understood by the tracer.  Contents of variable:\n" << var;
```
[the remark this answers is not in the document]
Comment (ezyang) [reply torch/csrc/jit/tracer.h]: On the contrary, while I was debugging this problem, as soon as I saw the contents of the variable I knew *exactly* what the problem was (it was zeros with a stripe of ones, and that screamed "inplace view update!") Since this error is a strictly compiler internal error, I would err on the side of more debug information.

[the remark this answers is not in the document]
Comment (apaszke) [reply torch/csrc/jit/tracer.h]: Does ATen only print summary of a tensor like PyTorch, or are these full contents? If the second, then I disagree. Having to scroll a few thousand lines up every time you’re trying to debug this error is not what you want (especially if you’re a user and hit this accidentally).

[the remark this answers is not in the document]
Comment (ezyang) [reply torch/csrc/jit/tracer.h]: If it is the second, that is surely an ATen bug to be fixed.

[the remark this answers is not in the document]
Comment (apaszke) [reply torch/csrc/jit/tracer.h]: Still, I don't want to see huge tensors in bug reports, because that doesn't carry any info for us. If you want to see what these tensors are because you don't know, you can always add the prints yourself.
